rust/src/ikev1/parser.rs

```diff
@@ -0,0 +1,732 @@
+                            },
+                            _ => None,
+                        },
+                        hex_value: match format.0 {
+                            0 => {
+                                if let Some(_variable_attribute_value) = variable_attribute_value {
+                                    Some(to_hex(_variable_attribute_value))
+                                } else {
+                                    None
+                                }
+                            }
+                            _ => None,
+                        }
+                    }
+                )
+            )
+        )
+    )
+}
+
+pub fn parse_nonce(
+    i: &[u8],
+    length: u16,
+) -> IResult<&[u8], NoncePayload> {
+    do_parse!(
+        i,
+        nonce_data: take!(length) >>
+        (
+            NoncePayload {
+                nonce_data: nonce_data
+            }
+        )
+    )
+}
+
+
+named! { pub parse_ikev1_payload_list<&[u8], Vec<IsakmpPayload>>,
+    many0!(
+        complete!(
+            do_parse!(
+                next_payload: be_u8 >>
+                reserved: be_u8 >>
+                payload_length: be_u16 >>
+                data: take!(payload_length - 4) >>
+                (
+                    IsakmpPayload {
+                        payload_header: IsakmpPayloadHeader {
+                            next_payload: next_payload,
+                            reserved: reserved,
+                            payload_length: payload_length
+                        },
+                        data: data
+                    }
+                )
+            )
+        )
+    )
+}
+
+#[derive(FromPrimitive, Debug)]
+pub enum IsakmpPayloadType {
+    None = 0,
+    SecurityAssociation = 1,
+    Proposal = 2,
+    Transform = 3,
+    KeyExchange = 4,
+    Identification = 5,
+    Certificate = 6,
+    CertificateRequest = 7,
+    Hash = 8,
+    Signature = 9,
+    Nonce = 10,
+    Notification = 11,
+    Delete = 12,
+    VendorID = 13,
+    SaKekPayload = 15,
+    SaTekPayload = 16,
+    KeyDownload = 17,
+    SequenceNumber = 18,
+    ProofOfPossession = 19,
+    NatDiscovery = 20,
+    NatOriginalAddress = 21,
+    GroupAssociatedPolicy = 22
+}
+
+impl fmt::Display for IsakmpPayloadType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub fn parse_payload<'a>(
+    payload_type: u8,
+    data: &'a[u8],
+    data_length: u16,
+    domain_of_interpretation: &mut Option<u32>,
+    key_exchange: &mut String,
+    nonce: &mut String,
+    transforms: &mut Vec<Vec<SaAttribute>>,
+    vendor_ids: &mut HashSet<String>,
+    payload_types: &mut HashSet<u8>) -> Result<(), ()>
+{
+    payload_types.insert(payload_type);
+
+    let element = num::FromPrimitive::from_u8(payload_type);
+    match element {
+        Some(IsakmpPayloadType::SecurityAssociation) => {
+            match parse_security_association(data, data_length) {
```

Comment:
I also suggest splitting some cases into subfunctions (`SecurityAssociation` and `Proposal`)

Do not mark subfunctions as `pub`, and the compiler will choose to inline them or not (rust is quite aggressive and usually inlines many functions)